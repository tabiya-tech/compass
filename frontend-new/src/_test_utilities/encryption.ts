import { webcrypto as crypto } from "crypto";
import { EncryptionConfig } from "src/sensitiveData/config/encryptionConfig";

// Constraining the key size to 2048 or 4096 bits for RSA as the backend expect the 4096 in calculating the maximum size of the encrypted key.
export async function generateRSACryptoPairKey(keySize: 2048 | 4096) {
  return await crypto.subtle.generateKey(
    {
      name: EncryptionConfig.RSA.ALGORITHM,
      modulusLength: keySize,
      // 3 is a common exponent in practice
      publicExponent: new Uint8Array([1, 0, 1]),
      hash: EncryptionConfig.RSA.KEY_HASH_FN,
    },
    true, // Whether the key is extractable (i.e. can be used in exportKey)
    ["encrypt", "decrypt"] // the key can be used for both encryption and decryption
  );
}

function arrayBufferToBase64(buffer: ArrayBuffer): string {
  // @ts-ignore
  const binary = String.fromCharCode(...new Uint8Array(buffer));

  return Buffer.from(binary, "binary").toString("base64");
}

/**
 * Export a CryptoKey to a PEM formatted string (public key)
 *
 * @param key The CryptoKey to export as a PEM formatted string
 * @returns The PEM formatted string.
 */
export async function exportCryptoPublicKey(key: CryptoKey) {
  // Export the key to ArrayBuffer
  const exported = await crypto.subtle.exportKey(EncryptionConfig.RSA.FORMAT, key);

  const base64String = arrayBufferToBase64(exported);

  // add the PEM header and footer to make it a PEM formatted string.
  return `-----BEGIN PUBLIC KEY-----\n${base64String}\n-----END PUBLIC KEY-----`;
}

/**
 * Decrypt the encrypted data with the private key
 * using RSA Algorithm.
 *
 * @param privateKey
 * @param encryptedData
 */
export async function decryptWithRSA(privateKey: CryptoKey, encryptedData: ArrayBuffer) {
  return await crypto.subtle.decrypt(
    {
      name: EncryptionConfig.RSA.ALGORITHM,
    },
    privateKey,
    encryptedData
  );
}

export async function decryptWithAES(encryptedData: ArrayBuffer, iv: ArrayBuffer, key: ArrayBuffer) {
  try {
    const importedKey = await crypto.subtle.importKey(
      EncryptionConfig.AES.FORMAT,
      key,
      EncryptionConfig.AES.ALGORITHM,
      true,
      ["decrypt"]
    );

    return await crypto.subtle.decrypt(
      {
        name: EncryptionConfig.AES.ALGORITHM,
        iv,
        length: EncryptionConfig.AES.KEY_LEN,
        tagLength: EncryptionConfig.AES.TAG_LEN,
      },
      importedKey,
      encryptedData
    );
  } catch (e) {
    console.error(e);
    throw e;
  }
}
