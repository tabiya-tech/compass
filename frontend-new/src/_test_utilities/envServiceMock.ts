jest.mock("src/envService", () => ({
  getFirebaseAPIKey: jest.fn(() => "mock-api-key"),
  getFirebaseDomain: jest.fn(() => "mock-auth-domain"),
  getBackendUrl: jest.fn(() => "mock-backend-url"),
  getApplicationLoginCode: jest.fn(() => ""),
  getApplicationRegistrationCode: jest.fn(() => ""),
  getLoginCodeDisabled: jest.fn(() => "false"),
  getRegistrationDisabled: jest.fn(() => "false"),
  getMetricsEnabled: jest.fn(() => "true"),
  getMetricsConfig: jest.fn(() => ""),
  getCvUploadEnabled: jest.fn(() => "true"),
  getSocialAuthDisabled: jest.fn(() => "false"),
}));
