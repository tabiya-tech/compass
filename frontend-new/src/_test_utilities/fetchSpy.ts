import { ExtendedRequestInit } from "src/utils/customFetch/customFetch";

export function setupFetchSpy(
  expectedStatus: number,
  expectedResponseBody: string | object | undefined,
  contentType: "" | "application/json;charset=UTF-8"
): jest.SpyInstance {
  const responseBody =
    typeof expectedResponseBody === "string" ? expectedResponseBody : JSON.stringify(expectedResponseBody);
  const expectedResponse = new Response(responseBody, {
    status: expectedStatus,
    headers: { "Content-Type": contentType },
  });
  return jest.spyOn(window, "fetch").mockResolvedValue(expectedResponse);
}

// spy on authenticated fetch requests
export function setupAPIServiceSpy(
  expectedStatus: number,
  expectedResponseBody: string | object | undefined,
  contentType: "" | "application/json;charset=UTF-8"
): jest.SpyInstance {
  const responseBody =
    typeof expectedResponseBody === "string" ? expectedResponseBody : JSON.stringify(expectedResponseBody);
  const expectedResponse = new Response(responseBody, {
    status: expectedStatus,
    headers: { "Content-Type": contentType },
  });

  jest.spyOn(window, "fetch").mockResolvedValue(expectedResponse);
  return jest.spyOn(require("src/utils/customFetch/customFetch"), "customFetch");
}

export function expectCorrectFetchRequest(fetchSpy: jest.SpyInstance, expectedUrl: string, expectedConfig: ExtendedRequestInit) {
  expect(fetchSpy).toHaveBeenCalledTimes(1);
  const [actualUrl, actualConfig] = fetchSpy.mock.calls[0];
  expect(actualUrl).toBe(expectedUrl);
  expect(actualConfig.method).toBe(expectedConfig.method);

  // Compare headers if they exist
  if(actualConfig.headers) {
    // Handle headers comparison - convert Headers object to plain object if needed
    const actualHeaders = actualConfig.headers instanceof Headers
      ? Object.fromEntries(actualConfig.headers.entries())
      : actualConfig.headers;
    expect(actualHeaders).toEqual(
      expect.objectContaining(expectedConfig.headers),
    );
  }

  // Compare bodies by parsing JSON if they exist
  if (actualConfig.body && expectedConfig.body) {
    let actualBody: unknown;
    let expectedBody: unknown;
    
    try {
      actualBody = JSON.parse(JSON.stringify(actualConfig.body));
      expectedBody = JSON.parse(JSON.stringify(expectedConfig.body));
    } catch (e) {
      console.error('Error parsing request bodies:', e);
      throw new Error('Failed to parse request bodies for comparison');
    }
    expect(actualBody).toEqual(expectedBody);
  }

  // Check any extra properties if provided
  type ExtendedConfigKey = keyof ExtendedRequestInit;
  const extraProperties: ExtendedConfigKey[] = [
    'expectedStatusCode',
    'serviceName',
    'serviceFunction',
    'failureMessage',
    'expectedContentType'
  ] as ExtendedConfigKey[];

  extraProperties.forEach((key) => {
    if (expectedConfig[key] !== undefined) {
      expect(actualConfig[key]).toEqual(expectedConfig[key]);
    }
  });
}

