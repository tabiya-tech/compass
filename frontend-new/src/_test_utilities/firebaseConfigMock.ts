import { AuthProviderIds } from "src/auth/auth.types";

const auth = jest.fn(() => ({
  signInWithCustomToken: jest.fn(),
  signInWithEmailAndPassword: jest.fn(),
  createUserWithEmailAndPassword: jest.fn(),
  signOut: jest.fn(),
  GoogleAuthProvider: { PROVIDER_ID: AuthProviderIds.GOOGLE },
}));

export function mockFirebaseConfig() {
  jest.spyOn(require("src/auth/firebaseConfig"), "auth").mockImplementation(auth);
}
