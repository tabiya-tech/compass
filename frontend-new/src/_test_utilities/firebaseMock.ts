import { AuthProviderIds } from "src/auth/auth.types";

const auth = {
  currentUser: {
    getIdToken: jest.fn(),
    getTokenResult: jest.fn(),
  },
  onAuthStateChanged: jest.fn().mockReturnValue(jest.fn()),
  signInWithPopup: jest.fn(),
  GoogleAuthProvider: { PROVIDER_ID: AuthProviderIds.GOOGLE },
};

const firebase = {
  initializeApp: jest.fn(),
  auth: () => auth,
};

export default firebase;
