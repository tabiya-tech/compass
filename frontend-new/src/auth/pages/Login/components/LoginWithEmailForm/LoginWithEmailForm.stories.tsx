import { Meta, StoryObj } from "@storybook/react";
import LoginWithEmailForm from "./LoginWithEmailForm";
import { action } from "@storybook/addon-actions";

const meta: Meta<typeof LoginWithEmailForm> = {
  title: "Auth/EmailAuth/LoginWithEmailForm",
  tags: ["autodocs"],
  component: LoginWithEmailForm,
};

export default meta;

export const Shown: StoryObj<typeof LoginWithEmailForm> = {
  args: {
    notifyOnEmailChanged: action("notifyOnEmailChanged"),
    notifyOnPasswordChanged: action("notifyOnPasswordChanged"),
    notifyOnFocused: action("notifyOnFocused"),
    isDisabled: false,
  },
};

export const LoggingIn: StoryObj<typeof LoginWithEmailForm> = {
  args: {
    notifyOnEmailChanged: action("notifyOnEmailChanged"),
    notifyOnPasswordChanged: action("notifyOnPasswordChanged"),
    notifyOnFocused: action("notifyOnFocused"),
    isDisabled: true,
  },
};
