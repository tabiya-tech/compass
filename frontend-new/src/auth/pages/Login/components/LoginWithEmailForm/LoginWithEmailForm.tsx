import { Box, TextField, useTheme } from "@mui/material";
import React from "react";

import PasswordInput from "src/theme/PasswordInput/PasswordInput";

const uniqueId = "8ab76120-a0d3-47b1-aac0-42d0169e0a58";

export const DATA_TEST_ID = {
  EMAIL_LOGIN_FORM_CONTAINER: `login-form-${uniqueId}`,
  EMAIL_LOGIN_FORM_EMAIL_INPUT: `login-email-input-${uniqueId}`,
  EMAIL_LOGIN_FORM_PASSWORD_INPUT: `login-password-input-${uniqueId}`,
};

export interface LoginFormProps {
  email: string;
  password: string;
  notifyOnEmailChanged: (email: string) => void;
  notifyOnPasswordChanged: (password: string) => void;
  notifyOnFocused: () => void;
  isDisabled: boolean;
}
const LoginWithEmailForm: React.FC<Readonly<LoginFormProps>> = ({
  email,
  password,
  notifyOnEmailChanged,
  notifyOnPasswordChanged,
  isDisabled,
}) => {
  const theme = useTheme();
  const handleEmailChange = (event: React.ChangeEvent<HTMLTextAreaElement | HTMLInputElement>) => {
    event.preventDefault();
    notifyOnEmailChanged(event.target.value);
  };
  const handlePasswordChange = (event: React.ChangeEvent<HTMLTextAreaElement | HTMLInputElement>) => {
    event.preventDefault();
    notifyOnPasswordChanged(event.target.value);
  };

  return (
    <Box
      display="flex"
      flexDirection="column"
      width={"100%"}
      justifyContent={"space-evenly"}
      gap={theme.fixedSpacing(theme.tabiyaSpacing.md)}
      data-testid={DATA_TEST_ID.EMAIL_LOGIN_FORM_CONTAINER}
    >
      <TextField
        fullWidth
        label="Email"
        type="email"
        variant="outlined"
        disabled={isDisabled}
        value={email}
        onChange={(e) => handleEmailChange(e)}
        inputProps={{ "data-testid": DATA_TEST_ID.EMAIL_LOGIN_FORM_EMAIL_INPUT }}
      />
      <PasswordInput
        fullWidth
        label="Password"
        variant="outlined"
        disabled={isDisabled}
        value={password}
        onChange={(e) => handlePasswordChange(e)}
        inputProps={{ "data-testid": DATA_TEST_ID.EMAIL_LOGIN_FORM_PASSWORD_INPUT }}
      />
    </Box>
  );
};

export default LoginWithEmailForm;
