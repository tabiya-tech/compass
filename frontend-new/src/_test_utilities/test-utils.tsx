// Based on https://testing-library.com/docs/react-testing-library/setup/
import React, { ReactElement, ReactNode } from "react";
import { render, renderHook, RenderHookOptions, RenderOptions } from "@testing-library/react";
import { ThemeProvider } from "@mui/material";
import applicationTheme, { ThemeMode } from "src/theme/applicationTheme/applicationTheme";

// Import the Firebase mock utilities
import "src/_test_utilities/firebaseMock";
import SnackbarProvider from "src/theme/SnackbarProvider/SnackbarProvider";
import { IsOnlineProvider } from "src/app/isOnlineProvider/IsOnlineProvider";
jest.mock("firebase/compat/app", () => {
  return {
    initializeApp: jest.fn(),
    auth: jest.fn().mockReturnValue({
      signOut: jest.fn(),
      currentUser: {
        getIdToken: jest.fn(),
        getTokenResult: jest.fn(),
      },
      onAuthStateChanged: jest.fn().mockReturnValue(jest.fn()),
      signInWithPopup: jest.fn(),
      GoogleAuthProvider: { PROVIDER_ID: "google.com" },
    }),
  };
});

jest.mock("firebaseui", () => {
  return {
    auth: {
      AuthUI: {
        start: jest.fn(),
        getInstance: jest.fn().mockReturnValue({
          start: jest.fn(),
          reset: jest.fn(),
        }),
        reset: jest.fn(),
      },
    },
  };
});

jest.mock("@sentry/react", () => ({
  withSentry: (Component: any) => Component,
  getFeedback: jest.fn(),
  withProfiler: (Component: any) => Component,
  wrapCreateBrowserRouter: (Component: any) => Component,
}));

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <IsOnlineProvider>
      <ThemeProvider theme={applicationTheme(ThemeMode.LIGHT)}>
        <SnackbarProvider>{children}</SnackbarProvider>
      </ThemeProvider>
    </IsOnlineProvider>
  );
};

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, "wrapper">) =>
  render(ui, { wrapper: AllTheProviders, ...options });

function customRenderHook<T>(
  hook: () => T,
  options?: Omit<RenderHookOptions<T>, "wrapper"> & { wrapper?: React.ComponentType }
) {
  const CombinedWrapper = ({ children }: { children: ReactNode }) => {
    const Wrapper = options?.wrapper ?? React.Fragment;
    return (
      <AllTheProviders>
        <Wrapper>{children}</Wrapper>
      </AllTheProviders>
    );
  };

  return renderHook(hook, { wrapper: CombinedWrapper, ...options });
}

export * from "@testing-library/react";
export { customRender as render };
export { customRenderHook as renderHook };
