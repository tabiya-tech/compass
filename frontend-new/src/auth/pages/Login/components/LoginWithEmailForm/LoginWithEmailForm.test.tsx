// silence chatty errors
import "src/_test_utilities/consoleMock";
import LoginWithEmailForm, { DATA_TEST_ID } from "./LoginWithEmailForm";
import { render, screen, fireEvent } from "src/_test_utilities/test-utils";
import { HashRouter } from "react-router-dom";
import React from "react";

describe("Testing LoginWithEmailForm component", () => {
  beforeEach(() => {
    // Clear console mocks and mock functions
    (console.error as jest.Mock).mockClear();
    (console.warn as jest.Mock).mockClear();
    jest.clearAllMocks();
  });

  test("it should show login form", async () => {
    // WHEN the component is rendered
    const givenNotifyOnEmailChange = jest.fn();
    const givenNotifyOnPasswordChange = jest.fn();
    const givenNotifyOnFocused = jest.fn();
    const givenEmailValue = "email";
    const givenPasswordValue = "password";
    const givenIsDisabled = false;

    render(
      <HashRouter>
        <LoginWithEmailForm
          email={givenEmailValue}
          password={givenPasswordValue}
          notifyOnEmailChanged={givenNotifyOnEmailChange}
          notifyOnPasswordChanged={givenNotifyOnPasswordChange}
          notifyOnFocused={givenNotifyOnFocused}
          isDisabled={givenIsDisabled}
        />
      </HashRouter>
    );

    // THEN expect no errors or warning to have occurred
    expect(console.error).not.toHaveBeenCalled();
    expect(console.warn).not.toHaveBeenCalled();

    // AND the email input should be rendered
    expect(screen.getByTestId(DATA_TEST_ID.EMAIL_LOGIN_FORM_EMAIL_INPUT)).toBeInTheDocument();

    // AND the password input should be rendered
    expect(screen.getByTestId(DATA_TEST_ID.EMAIL_LOGIN_FORM_PASSWORD_INPUT)).toBeInTheDocument();

    // AND the component should match the snapshot
    expect(screen.getByTestId(DATA_TEST_ID.EMAIL_LOGIN_FORM_CONTAINER)).toMatchSnapshot();
  });

  describe("action tests", () => {
    test("should call notigyOnEmailChanged when the email field changes", async () => {
      // WHEN the component is rendered
      const givenNotifyOnEmailChange = jest.fn();
      const givenNotifyOnPasswordChange = jest.fn();
      const givenNotifyOnFocused = jest.fn();
      const givenEmailValue = "email";
      const givenPasswordValue = "password";
      const givenIsDisabled = false;

      render(
        <HashRouter>
          <LoginWithEmailForm
            email={givenEmailValue}
            password={givenPasswordValue}
            notifyOnEmailChanged={givenNotifyOnEmailChange}
            notifyOnPasswordChanged={givenNotifyOnPasswordChange}
            notifyOnFocused={givenNotifyOnFocused}
            isDisabled={givenIsDisabled}
          />
        </HashRouter>
      );

      // Simulate form input and submission
      fireEvent.change(screen.getByTestId(DATA_TEST_ID.EMAIL_LOGIN_FORM_EMAIL_INPUT), {
        target: { value: "foo@bar.baz" },
      });

      // THEN expect the notifyOnEmailChanged function to have been called
      expect(givenNotifyOnEmailChange).toHaveBeenCalledWith("foo@bar.baz");
    });

    test("should call notifyOnPasswordChanged when the password field changes", async () => {
      // WHEN the component is rendered
      const givenNotifyOnEmailChange = jest.fn();
      const givenNotifyOnPasswordChange = jest.fn();
      const givenNotifyOnFocused = jest.fn();
      const givenEmailValue = "email";
      const givenPasswordValue = "password";
      const givenIsDisabled = false;

      render(
        <HashRouter>
          <LoginWithEmailForm
            email={givenEmailValue}
            password={givenPasswordValue}
            notifyOnEmailChanged={givenNotifyOnEmailChange}
            notifyOnPasswordChanged={givenNotifyOnPasswordChange}
            notifyOnFocused={givenNotifyOnFocused}
            isDisabled={givenIsDisabled}
          />
        </HashRouter>
      );

      // Simulate form input and submission
      fireEvent.change(screen.getByTestId(DATA_TEST_ID.EMAIL_LOGIN_FORM_PASSWORD_INPUT), {
        target: { value: "Pa$$word123" },
      });

      // THEN expect the notifyOnPasswordChanged function to have been called
      expect(givenNotifyOnPasswordChange).toHaveBeenCalledWith("Pa$$word123");
    });
  });
});
