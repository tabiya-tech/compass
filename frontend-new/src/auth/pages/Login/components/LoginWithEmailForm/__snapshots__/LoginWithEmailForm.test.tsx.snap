// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing LoginWithEmailForm component it should show login form 1`] = `
<div
  class="MuiBox-root css-6qz4l6"
  data-testid="login-form-8ab76120-a0d3-47b1-aac0-42d0169e0a58"
>
  <div
    class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-1pcwqdd-MuiFormControl-root-MuiTextField-root"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1d7vtq4-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="true"
      for=":r0:"
      id=":r0:-label"
    >
      Email
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-1vdto9g-MuiInputBase-root-MuiOutlinedInput-root"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiOutlinedInput-input css-kqqmrr-MuiInputBase-input-MuiOutlinedInput-input"
        data-testid="login-email-input-8ab76120-a0d3-47b1-aac0-42d0169e0a58"
        id=":r0:"
        type="email"
        value="email"
      />
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-14lo706"
        >
          <span>
            Email
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
  <div
    class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-1pcwqdd-MuiFormControl-root-MuiTextField-root"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1d7vtq4-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="true"
      for=":r1:"
      id=":r1:-label"
    >
      Password
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd css-1bk0u5a-MuiInputBase-root-MuiOutlinedInput-root"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-hjin93-MuiInputBase-input-MuiOutlinedInput-input"
        data-testid="login-password-input-8ab76120-a0d3-47b1-aac0-42d0169e0a58"
        id=":r1:"
        type="password"
        value="password"
      />
      <div
        class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
      >
        <button
          aria-label="toggle password visibility"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-hli2kd-MuiButtonBase-root-MuiIconButton-root"
          data-testid="b7499b01-8082-4209-8667-c7d559a70cag-icon-button"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-etbowg-MuiSvgIcon-root"
            data-testid="b7499b01-8082-4209-8667-c7d559a70cag-visibility-on-icon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5M12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5m0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-14lo706"
        >
          <span>
            Password
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;
